Qute - introduce template fragments (#28216)
quarkusio/quarkus: docs/src/main/asciidoc/qute-reference.adoc

</ol>
{/fragment}
----
<1> Defines a fragment with identifier `item_aliases`. Note that only alphanumeric characters and underscores can be used in the identifier.

You can obtain a fragment programatically via the `io.quarkus.qute.Template.getFragment(String)` method.

.Obtaining a Fragment
[source,java]
----
@Inject
Template item;

String useTheFragment() {
   return item.getFragment("item_aliases") <1>
            .data("aliases", List.of("Foo","Bar")) <2>
            .render();
}
----
<1> Obtains the template fragment with identifier `item_aliases`.
<2> Make sure the data are set correctly.

The snippet above should render something like:

[source,html]
----
<h2>Aliases</h2>
<ol>
    <li>Foo</li>
    <li>Bar</li>
</ol>
----

You can also include a template with an `{#include}` section inside another template or the template that defines the fragment.

.Including a Fragment in `user.html`
[source,html]
----
<h1>User - {user.name}</h1>

<p>This document contains a detailed info about a user.</p>

{#include item[item_aliases] aliases=user.aliases /} <1><2>
----
<1> The `item[item_aliases]` parameter is translated as: _use the fragment `item_aliases` from the template `item`._

> <2> The `aliases` parameter is used to pass the relevant data.

Reviewer: I still don't understand what `aliases` does here, mind to elaborate a bit more?

Author: The `aliases` expression is used in the `item_aliases` fragment in the `{#for alias in aliases}` section. In this particular case, we use the expression `user.aliases` as the value. Let me try to describe it a little bit better...

===== Hidden Fragments

By default, a fragment is normally rendered as a part of the original template.
However, sometimes it might be useful to mark a fragment as _hidden_ with `rendered=false`.
An interesting use case would be a fragment that can be used multiple-times inside the template that defines it.

.Fragment Definition in `item.html`
[source,html]
----

> {#fragment id=strong rendered=false} <1>

Reviewer: Can one specify an expression in the rendered attribute?

Author: Yes, that's possible. Let's improve the description...

<strong>{val}</strong>
{/fragment}

<h1>My page</h1>
<p>This document 
{#include [strong] val='contains' /} <2>
a lot of 
{#include [strong] val='information' /} <3>
!</p>
----

> <1> Defines an invisible fragment with identifier `strong`.

Reviewer: ```suggestion
<1> Defines a hidden fragment with identifier `strong`.
```

Maybe stick to a common term to avoid confusion
